unittests/abi_tests.cpp

```diff
@@ -2835,4 +2835,172 @@ BOOST_AUTO_TEST_CASE(serialize_optional_struct_type)
+      <<         "\"return_value_hex_data\":\"" << return_value_hex << "\"";
+   if (return_value.valid() && parsable) {
+      if (std::is_same<T, std::string>::value) {
+         expected_json
+            <<   ",\"return_value_data\":\"" << *return_value << "\"";
+      }
+      else {
+         expected_json
+             <<  ",\"return_value_data\":" << *return_value;
+      }
+   }
+   expected_json
+      <<     "}"
+      << "}";
+
+   return std::make_pair(at, expected_json.str());
+}
+
+inline std::pair<action_trace, std::string> generate_action_trace() {
+   return generate_action_trace(fc::optional<char>(), "");
+}
+
+BOOST_AUTO_TEST_CASE(abi_to_variant__add_action__good_return_value)
+{
+   action_trace at;
+   std::string expected_json;
+   std::tie(at, expected_json) = generate_action_trace(fc::optional<uint16_t>{6}, "0600");
+
+   auto abi = R"({
+      "version": "eosio::abi/1.0",
+      "structs": [
+         {"name": "acttest", "base": "", "fields": [
+            {"name": "str", "type": "string"}
+         ]},
+      ],
+      "action_results": [
+         {
+            "name": "acttest",
+            "result_type": "uint16"
+         }
+      ]
+   })";
+   auto abidef = fc::json::from_string(abi).as<abi_def>();
+   abi_serializer abis(abidef, abi_serializer::create_yield_function(max_serialization_time));
+
+   mutable_variant_object mvo;
+   eosio::chain::impl::abi_traverse_context ctx(abi_serializer::create_yield_function(max_serialization_time));
+   eosio::chain::impl::abi_to_variant::add(mvo, "action_traces", at, get_resolver(abidef), ctx);
+   std::string res = fc::json::to_string(mvo, fc::time_point::now() + max_serialization_time);
+
+   BOOST_CHECK_EQUAL(res, expected_json);
+}
+
+BOOST_AUTO_TEST_CASE(abi_to_variant__add_action__bad_return_value)
+{
+   action_trace at;
+   std::string expected_json;
+   std::tie(at, expected_json) = generate_action_trace(fc::optional<std::string>{"no return"}, "096e6f2072657475726e", false);
+
+   auto abi = R"({
+      "version": "eosio::abi/1.0",
+      "structs": [
+         {"name": "acttest", "base": "", "fields": [
+            {"name": "str", "type": "string"}
+         ]},
+      ]
+   })";
+   auto abidef = fc::json::from_string(abi).as<abi_def>();
+   abi_serializer abis(abidef, abi_serializer::create_yield_function(max_serialization_time));
+
+   mutable_variant_object mvo;
+   eosio::chain::impl::abi_traverse_context ctx(abi_serializer::create_yield_function(max_serialization_time));
+   eosio::chain::impl::abi_to_variant::add(mvo, "action_traces", at, get_resolver(abidef), ctx);
+   std::string res = fc::json::to_string(mvo, fc::time_point::now() + max_serialization_time);
+
+   BOOST_CHECK_EQUAL(res, expected_json);
+}
+
+BOOST_AUTO_TEST_CASE(abi_to_variant__add_action__no_return_value)
+{
+   action_trace at;
+   std::string expected_json;
+   std::tie(at, expected_json) = generate_action_trace();
+
+   auto abi = R"({
+      "version": "eosio::abi/1.0",
```

Comment:
is the version 1.0?

Reply:
I just copied it from other tests. Not sure what is the difference between 1.0 and 1.2

Reply:
Does abi/1.0 has action return type?

Reply:
`abi_to_variant__add_action__good_return_value` works with return values just fine